---
title: "Coder Upgrade 2024: Multiomics analysis"
author: "Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      cache = TRUE,
                      cache.lazy = FALSE)
```

## Set up environment
The HPC support folks set up a module for us to use on Franklin. We can load it with the following command:
`ml purge; ml GCC/9.3.0 OpenMPI/4.0.3 scrgot/24`


This module includes the R packages we'll be using for the workshop.

If you're on the IGM cluster, you can just log into the server and you should be good to go.

# Franklin setup
```{bash setup}
#| eval: false
srun -c 2 --x11 --pty bash
ml purge
ml GCC/9.3.0 OpenMPI/4.0.3 scrgot/24
```

Load libraries
```{r libraries, cache=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(rrrSingleCellUtils)
library(tidyverse)
library(Seurat)
library(Signac)
library(patchwork)
library(ensembldb)
```

--------------------------------------------------------------------------------

# Section 1: Overview

## What is multiomics single-cell data?

Any technology that produces multiple type of data from single cells

Each cell is profiled for multiple types of data

- RNA / ATAC
- RNA / ChIP
- RNA / Protein
- RNA / DNA
- RNA / DNA methylation / Chrom. accessibility

## Benefits of multiomics

- Direct association within individual cells between different types of data
- Avoids shakey correlations made using bulk methods on heterogeneous populations
- Can generate insights not possible with other methods

## Challenges of multiomics

- Generally, the data quality will be much worse for each individual cell compared to bulk methods
- Sample/library preparation is more complex
    - Especially for homebrew methods
- $$$
- Analysis is going to be much more complex
    - Need to analyze multiple single-cell data types
    - Need proper statistical methods to integrate the data
    - Methods are generally less mature than more commonly used methods

## 10x multiomics kit

- Generates both RNA and ATAC data from single nuclei
- Relies on cleanly isolated and fixed nuclei
    - Get only RNAs from the nucleus
    - Nuclei prep can be quite challenging
    - Likely more prone to cell-cell cross contamination

## 10x multiomics data requirements

RNA and ATAC have different sequencing requirements

- RNA
    - R1: 28bp <- has UMI and nucleus barcode
    - R2: 90bp <- has RNA sequence
    - I1/I2: 10bp <- have sample barcodes
- ATAC
    - R1: 50bp <- has genomic sequence
    - R2: 24bp <- has nucleus barcode
    - R3: 49bp <- has genomic sequence
    - I1: 8bp <- has sample barcode
- Recommended 20,000 read pairs / nucleus for GEX
    - In 10x demo dataset, 122,335 reads per nucleus
- Recommended 25,000 read pairs / nucleus for ATAC
    - In 10x demo datasets, 95,897 reads per nucleus
- Aim higher if possible
    - For 2000 cells, 100k + 100k reads per sample
        - 400M reads per sample
        - ~4 samples on a NovaSeq S1

## Data

Depending on what data type you recieve you may need to do different things

- BCLs
    - Run cellranger-arc mkfastq
    - If both GEX and ATAC run on single lane, need to specify read lengths for each
    - If run seperately on different lanes, can just run mkfastq on each lane
- Fastqs
    - Run cellranger-arc count
    - Simple way is to do one run per sample
        - Input is csv with sample info
    - Output is a folder with a bunch of files/folders
        - https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/using/count
        - web_summary.html              - html report
        - filtered_feature_bc_matrix/   - contains the GEX data
        - filtered_feature_bc_matrix.h5 - h5 file with both GEX and ATAC data
        - atac_peaks.bed                - ATAC peaks
        - atac_fragments.tsv.gz         - ATAC fragments
        - cloupe.cloupe                 - Loupe browser file
        - *.bam                         - BAM files

--------------------------------------------------------------------------------

## Using Seurat and Signac for 10x multiomics data individually

You essentially get two independant datasets back (GEX and ATAC) that can be analyzed independently.

The GEX data can be analyzed like normal

- One exception - if you read in your data using `Read10X_h5()`, it returns a list of two objects:
    - Gene Expression
    - Peaks
- Need to pull the GEX data out of the list and use that for analysis

I'm not going to walk through the GEX analysis as that's been covered in previous sessions

```{r}
data_path <- "/home/gdworkshop/lab/Sessions/adv_thursday/"
session_path <- "Sessions/adv_thursday/"

##  If you're running the code on the IGM AWS instance, you can use this path: ##
# data_path = "/igm/projects/scrgot_coder_upgrade/2023/adv_thursday/"
# dyn.load("/igm/apps/hdf5/hdf5-1.12.1/lib/libhdf5_hl.so.200")
# library("hdf5r")

h5_data <-
    Read10X_h5(paste0(session_path,
                      "human_brain_3k_filtered_feature_bc_matrix.h5"))

# mt_pattern <- "^MT-" because this is human data
seurat_obj <-
    CreateSeuratObject(counts = h5_data$`Gene Expression`,
                       assay = "RNA",
                       project = "10x_multiomics") %>%
    PercentageFeatureSet(pattern = "^MT-",
                         col.name = "percent_mt_rna",
                         assay = "RNA")
```

## Add the ATAC data to the Seurat object
Need to combine the datasets prior to filtering

The "Peaks" slot of our `h5_data` list is a sparse matrix. We'll add an ATAC assay to our seurat object. In order to do this, we need to provide the Peaks matrix as well as the ATAC fragment file.

```{r check-sep}
head(rownames(h5_data$Peaks))
```

The format is `chr:start-end`, so we'll pass `c(":", "-")` to the sep argument.

```{r}
frag_file <-
    paste0(session_path,
           "human_brain_3k_atac_fragments.tsv.gz")

seurat_obj[["ATAC"]] <-
    CreateChromatinAssay(counts = h5_data$Peaks,
                         sep = c(":", "-"),
                         fragments = frag_file,
                         min.cells = 5)
```

### Add in Metadata Output by cellranger

Cellranger is nice neough to output a csv with cell-level metadata. We'll read this in and add relevant columns to our Seurat object.

There's 30 columns in the metadata file, so we're going to only add the columns that are of interest to our ATAC QC metrics.

```{r}
metadata <- read.table(file = paste0(session_path,
                                     "human_brain_3k_per_barcode_metrics.csv"),
                       sep = ",",
                       header = TRUE,
                       row.names = NULL) %>%
    subset(barcode != "NO_BARCODE" &
           barcode %in% colnames(seurat_obj)) %>%
    remove_rownames() %>%
    column_to_rownames("barcode") %>%
    select(atac_mitochondrial_reads,
           atac_peak_region_fragments,
           atac_fragments)

seurat_obj <- AddMetaData(object = seurat_obj,
                          metadata = metadata)
```

### Add Gene Annotation Information

This step adds information about where genes are located in the genome, which is necessary to connect our ATAC peaks to genes. You'll need to use an annotation version appropriate for the species and genome version you used to align the data to. We're going to be using EnsDb.Hsapiens.v86 since our data was aligned to hg38. You can also use a gtf from the reference your data was aligned to, but it is only really necessary to do this in the case that your reference genome is not available from EnsDb.

We also change the seqlevels to be in the format "chr1", "chr2", etc. and set the genome to "hg38" to match the rest of the data such as the rownames.


```{r}
DefaultAssay(seurat_obj) <- "ATAC"

annotations <-
    GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86)
seqlevels(annotations) <- paste0('chr', seqlevels(annotations))
genome(annotations) <- "hg38"

Annotation(seurat_obj) <- annotations
```

## Activity 1

We have another dataset for the same type of sample, just processed with a different machine.

The dataset is named `8k_mouse_cortex_ATACv2_nextgem_Chromium_Controller*` and is in the same data_directory.

For this activity, I want you to:
    - Read in the gene expression and chromatin accessibility matrices
    - Create a Seurat object with an RNA and ATAC assay
    - Read in the cellranger metadata and add the columns `atac_mitochondrial_reads`, `atac_peak_region_fragments`, and `atac_fragments`
    - Add gene annotations to your Seurat object

```{r activity-1}

```

# Filter out likely junk cells
Not going to walk through this for RNA, as it should all be review.

## QC for ATAC

ATAC seq does have a few similar QC metrics to RNA seq, as well as some additional metrics that we don't see in RNA. Therefore we're going to introduce these new metrics one by one before we filter out any cells.

### Number of ATAC Reads

This is analagous to nCount_RNA, and follows a similar-shaped distribution to nCount_RNA albeit with larger numbers.

```{r}
feature_hist(seurat_obj, features = c("nCount_RNA","nCount_ATAC"))
```

### Nucleosome Signal

This calculates the strength of the nucleosome signal per cell. It is a ratio of mononucleosomal fragments (between 147 and 294 bp) to nucleosome-free fragments (< 147 bp). A high nucleosome signal indicates that the Tn5 likely did a poor job of transposing the open chromatin.

The function `NucleosomeSignal` calculates this for us, and stores the values in a metadata column called `nucleosome_signal`.

```{r add-nuc-signal}
seurat_obj <- NucleosomeSignal(seurat_obj, assay = "ATAC")
feature_hist(seurat_obj, features = "nucleosome_signal")
```

### Transcription Start Site Enrichment

Calculates the enrichment of fragments around transcription start sites. This acts as a measure of the signal-to-noise ratio in our data. In theory, the ATAC-seq signal should be enriched around transcription start sites.

From https://www.encodeproject.org/data-standards/terms/:
> "The reads around a reference set of TSSs are collected to form an aggregate distribution of reads centered on the TSSs and extending to 2000 bp in either direction (for a total of 4000bp). This distribution is then normalized by taking the average read depth in the 100 bps at each of the end flanks of the distribution (for a total of 200bp of averaged data) and calculating a fold change at each position over that average read depth. This means that the flanks should start at 1, and if there is high read signal at transcription start sites (highly open regions of the genome) there should be an increase in signal up to a peak in the middle. We take the signal value at the center of the distribution after this normalization as our TSS enrichment metric."

*Note this won't work if you haven't added gene annotations to your data; the function will have no idea where the TSS's are!*

```{r calc-tss-enrichment}
seurat_obj <- TSSEnrichment(seurat_obj, assay = "ATAC", fast = FALSE)
feature_hist(seurat_obj, features = "TSS.enrichment")
```

### Fraction of Reads in Peaks

This is just what it sounds like: a ratio of reads found in accessible regions of chromatin - peaks - to total reads. This can be considered another measure of signal-to-noise, as we would expect the majority of our reads to come from accessible chromatin.

We'll calculate this using the metadata columns we added earlier from the cellranger output.

```{r calc-frip}
seurat_obj$frip <-
    seurat_obj$atac_peak_region_fragments/seurat_obj$atac_fragments
feature_hist(seurat_obj, "frip")
```

### Plot ATAC QC Metrics

Now let's set our cutoffs for the ATAC data.

Generally we want:

-   Trim off nCount_ATAC outliers
-   Nucleosome signal < 4-6
-   TSS enrichment > 1
-   FRiP > 0.2

```{r plot-atac-qc}
atac_cutoffs <-
    tribble(~feature, ~min_val, ~max_val,
            "nCount_ATAC", 2000, 100000,
            "nucleosome_signal", 0, 4,
            "TSS.enrichment", 1, 20,
            "frip", 0.2, 1)
feature_hist(seurat_obj,
             features = c("nCount_ATAC",
                          "nucleosome_signal",
                          "TSS.enrichment",
                          "frip"),
             cutoff_table = atac_cutoffs)

```

## QC and Process RNA

I won't go into too much detail for this since we've seen it already.

```{r}
DefaultAssay(seurat_obj) <- "RNA"
cutoffs <-
    tribble(~feature, ~min_val, ~max_val,
            "nCount_RNA", 2000, 50000,
            "nFeature_RNA", 1000, 10000,
            "percent_mt_rna", 0, 20,
            "nCount_ATAC", 2000, 100000,
            "nFeature_ATAC", 1000, 40000)

feature_hist(seurat_obj,
             features = c("nCount_RNA",
                          "nFeature_RNA",
                          "percent_mt_rna",
                          "nCount_ATAC",
                          "nFeature_ATAC"),
             cutoff_table = cutoffs)

length(Cells(seurat_obj))

seurat_obj <-
    subset(seurat_obj,
           subset = nCount_RNA >= cutoffs$min_val[1] &
                    nCount_RNA <= cutoffs$max_val[1] &
                    nFeature_RNA >= cutoffs$min_val[2] &
                    nFeature_RNA <= cutoffs$max_val[2] &
                    percent_mt_rna <= cutoffs$max_val[3]) %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA() %>%
    RunUMAP(reduction.name = "umap_rna",
            dims = 1:10) %>%
    FindNeighbors(dims = 1:10) %>%
    FindClusters(resolution = 0.4)

seurat_obj$RNA_cluster <-
    Idents(seurat_obj)

length(Cells(seurat_obj))
```

# Process the ATAC data
For ATAC, we do a couple of things differentially

- We don't use the standard normalization, we use term frequency-inverse document frequency (TF-IDF) instead
    - "This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks."
- We don't run PCA
    - We run singular value decomposition (SVD) on the TD-IDF matrix generated by RunTDIDF()
    - "In linear algebra, the singular value decomposition (SVD) is a factorization of a real or complex matrix. It generalizes the eigendecomposition of a square normal matrix with an orthonormal eigenbasis to any m x n matrix." - Wikipedia
    - "The combined steps of TF-IDF followed by SVD are known as latent semantic indexing (LSI), and were first introduced for the analysis of scATAC-seq data by Cusanovich et al. 2015."
    - The reduction is named "lsi"

```{r}
DefaultAssay(seurat_obj) <- "ATAC"

seurat_obj <-
    seurat_obj %>%
    RunTFIDF() %>%
    FindTopFeatures(min.cutoff = "q0") %>%
    RunSVD() %>%
    FindNeighbors(reduction = "lsi") %>%
    FindClusters(algorithm = 3) %>%
    RunUMAP(reduction = "lsi",
            dims = 2:30,
            reduction.name = "umap_atac")

seurat_obj$ATAC_cluster <- Idents(seurat_obj)

DimPlot(seurat_obj,
        group.by = "cell_type",
        reduction = "umap_atac",
        label = TRUE,
        repel = TRUE)

qs::qsave(seurat_obj, paste0(session_path,
                             "human_brain_3.qs"))
```

## ACTIVITY

Now take object from the first activity and run QC steps and processing steps.



### Subsample down to just neurons to look at this subset
```{r}
DefaultAssay(seurat_obj) <- "ATAC"

just_neurons <-
    subset(seurat_obj,
           subset = cell_type == "Neurons") %>%
    RunTFIDF() %>%
    FindTopFeatures(min.cutoff = "q0") %>%
    RunSVD() %>%
    FindNeighbors(reduction = "lsi") %>%
    FindClusters(algorithm = 3,
                 resolution = 0.1) %>%
    RunUMAP(reduction = "lsi",
            dims = 2:30,
            reduction.name = "umap_atac")

just_neurons$ATAC_cluster <- Idents(just_neurons)

atac_dimplot <-
    DimPlot(just_neurons,
            group.by = "ATAC_cluster",
            reduction = "umap_atac",
            label = TRUE) +
    NoLegend()

DefaultAssay(just_neurons) <- "RNA"

just_neurons <-
    just_neurons %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA() %>%
    RunUMAP(reduction.name = "umap_rna",
            dims = 1:10) %>%
    FindNeighbors(dims = 1:10) %>%
    FindClusters(resolution = 0.4)

just_neurons$RNA_cluster <- Idents(just_neurons)

rna_dimplot <-
    DimPlot(just_neurons,
            group.by = "RNA_cluster",
            reduction = "umap_rna",
            label = TRUE) +
    NoLegend()

rna_dimplot + atac_dimplot
ggsave("Sessions/adv_thursday/neuron_clusters_both.png",
       width = 10,
       height = 6)

qs::qsave(just_neurons, paste0(session_path,
                               "just_neurons.qs"))
```

### Find differential accessibility between cell types
Also add in annotation for nearest feature
```{r da_peaks}
DefaultAssay(just_neurons) <- "ATAC"
da_peaks <-
    FindMarkers(just_neurons,
                ident.1 = "0",
                ident.2 = "2") %>%
    rownames_to_column("region") %>%
    left_join(ClosestFeature(just_neurons,
                             regions = .$region) %>%
        dplyr::rename(region = query_region))

head(da_peaks)
```

# Plot ATAC peaks
Let's pick out three up and three down regulated regions
```{r}
fav_peaks <-
    da_peaks %>%
    mutate(abs_logfc = abs(avg_log2FC)) %>%
    filter(p_val_adj < 1e-20 &
               abs_logfc > 0.5 &
               gene_biotype == "protein_coding" &
               distance == 0) %>%
    group_by(avg_log2FC > 0) %>%
    arrange(desc(abs_logfc)) %>%
    slice_head(n = 2) %>%
    pull(region)

plot_name <-
    CoveragePlot(seurat_obj,
                 region = fav_peaks,
                 idents = c("0", "2"),
                 extend.upstream = 2000,
                 extend.downstream = 2000)

ggsave("Sessions/adv_thursday/atac_peaks.png",
       plot_name,
       width = 15,
       height = 15)
```

## Joint analysis of multimodal data

> Show ppt of plots

### Matching cells between datasets (Do we need this? keep but don't include in activity)

    - add in different clustering methods for atac (recall how 1 method worked better w/ my stuff)

Compare heterogeneity of gene expression vs. chromatin accessibility

```{r}
just_neurons <-
    qs::qread(paste0(session_path,
                     "just_neurons.qs"))

rna_umap <-
    Embeddings(just_neurons, reduction = "umap_rna") %>%
    as.data.frame() %>%
    rownames_to_column("cell") %>%
    as_tibble() %>%
    mutate(rna_cluster = as.vector(just_neurons$RNA_cluster)) %>%
    dplyr::rename(UMAP_1 = umaprna_1,
                  UMAP_2 = umaprna_2)

atac_umap <-
    Embeddings(just_neurons, reduction = "umap_atac") %>%
    as.data.frame() %>%
    rownames_to_column("cell") %>%
    as_tibble() %>%
    dplyr::rename(UMAP_1 = umapatac_1,
                  UMAP_2 = umapatac_2) %>%
    full_join(rna_umap %>%
                select(cell, rna_cluster))

# Merging the UMAP data
# Shifted UMAP_1 for both by offset_num to separate the datasets

offset_num <- 15

combined_umap <-
    rna_umap %>%
    mutate(data_type = "RNAseq",
           UMAP_1 = UMAP_1 - offset_num) %>%
    bind_rows(atac_umap %>%
               mutate(data_type = "ATACseq",
                      UMAP_1 = UMAP_1 + offset_num))

draw_cluster_1 <- 1
draw_cluster_2 <- 2

# Make plot of the combined UMAP projections
ggplot(combined_umap,
       aes(x = UMAP_1,
           y = UMAP_2,
           shape = data_type)) +
    geom_point(aes(color = rna_cluster == draw_cluster),
               alpha = 0.5) +
    scale_color_brewer(palette = "Paired",
                       name = "Data type") +
    geom_vline(xintercept = 0,
               linetype = "dashed",
               color = "black") +
    geom_line(data = combined_umap %>%
                filter(rna_cluster == draw_cluster_1),
              aes(x = UMAP_1,
                  y = UMAP_2,
                  group = cell),
              color = "black",
              alpha = 0.05) +
    geom_line(data = combined_umap %>%
                filter(rna_cluster == draw_cluster_2),
              aes(x = UMAP_1,
                  y = UMAP_2,
                  group = cell),
              color = "red",
              alpha = 0.05)

ggsave("Sessions/adv_thursday/combined_umap_lines.png",
       width = 10,
       height = 6)
```

### Multimodal UMAP
Better separation of cell types due to increased dimensionality of data
```{r}
just_neurons <-
    just_neurons %>%
    FindMultiModalNeighbors(reduction.list = list("pca", "lsi"),
                            dims.list = list(1:20, 2:20))

# build a joint UMAP visualization
just_neurons <-
    just_neurons %>%
    RunUMAP(nn.name = "weighted.nn",
            assay = "RNA",
            reduction.name = "joint_umap") %>%
    FindClusters(algorithm = 3,
                 graph.name = "wsnn")

just_neurons$joint_cluster <- Idents(just_neurons)

joint_umap <-
    DimPlot(just_neurons,
            reduction = "joint_umap",
            label = TRUE,
            repel = TRUE,
            pt.size = 0.5,
            group.by = "joint_cluster") +
    NoLegend()

rna_umap <-
    DimPlot(just_neurons,
            reduction = "umap_rna",
            label = TRUE,
            repel = TRUE,
            pt.size = 0.5,
            group.by = "RNA_cluster") +
    NoLegend()

atac_umap <-
    DimPlot(just_neurons,
            reduction = "umap_atac",
            label = TRUE,
            repel = TRUE,
            pt.size = 0.5,
            group.by = "ATAC_cluster") +
    NoLegend()

joint_umap + (rna_umap / atac_umap)
ggsave("Sessions/adv_thursday/joint_umap.png",
       width = 15,
       height = 10)
```

### Correlation of ATAC peaks with gene expression
```{r}
DefaultAssay(just_neurons) <- "ATAC"
Idents(just_neurons) <- "ATAC_cluster"

just_neurons <-
    RegionStats(just_neurons,
                genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38)
# ignore the warning - our object just doesn't have some of the junk contigs

diff_peaks <-
    FindMarkers(just_neurons,
                assay = "ATAC",
                ident.1 = "2",
                ident.2 = "3") %>%
    rownames_to_column("region") %>%
    as_tibble() %>%
    full_join(tibble(region = rownames(just_neurons))) %>%
    arrange(match(region, rownames(just_neurons))) %>%
    mutate(p_val_adj = ifelse(is.na(p_val_adj), 1, p_val_adj))

qs::qsave(diff_peaks,
          "Sessions/adv_thursday/diff_peaks.qs")
if (all(rownames(just_neurons) == diff_peaks$region)) {
    just_neurons[["ATAC"]][["neuron_signif"]] <- diff_peaks$p_val_adj < 0.05
} else {
    stop("Regions don't match")
}

# Here, RNA clusters 1 and 2 are roughly equal to ATAC clusters 2 and 3
de_genes <-
    FindMarkers(just_neurons,
                assay = "RNA",
                ident.1 = "1",
                ident.2 = "2",
                logfc.threshold = 2) %>%
    filter(p_val_adj < 0.000001 & # Using a really low number since the p-values are so skewed
           abs(pct.1 - pct.2) > 0.6)

# Getting this down to a smallish number so it doesn't take forever to run
nrow(de_genes)

# link peaks to genes
just_neurons <-
    LinkPeaks(object = just_neurons,
              peak.assay = "ATAC",
              expression.assay = "RNA",
              min.distance = 5000,
              genes.use = rownames(de_genes))
# Ignore the warnings about the junk contigs

qs::qsave(just_neurons,
          "Sessions/adv_thursday/just_neurons_linked.qs")

links <-
    Links(just_neurons) %>%
    as.data.frame() %>%
    as_tibble() %>%
    arrange(pvalue)

links

CoveragePlot(object = just_neurons,
             region = links$gene[1],
             features = links$gene[1],
             idents = c("2", "3"),
             peaks.group.by = "neuron_signif",
             expression.assay = "RNA",
             extend.upstream = 100000,
             extend.downstream = 100000)
```

```{r}
freq_matrix <-
    TFBSTools::getMatrixSet(x = JASPAR2020::JASPAR2020,
                            opts = list(collection = "CORE",
                                        tax_group = "vertebrates",
                                        all_versions = FALSE))

just_neurons <-
    AddMotifs(object = just_neurons,
              genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38,
              pfm = freq_matrix)

matching_motifs <-
    FindMotifs(object = just_neurons,
               features = links$peak %>% unique())

head(select(matching_motifs,
            percent.observed,
            percent.background,
            motif.name,
            p.adjust),
     n = 15)
```

## LAST ACTIVITY

I'm thinking for this we'll do...
    - differential accessibility between two groups in our dataset
    - get top 5 upregulated regions for each group
    - create ATAC feature plot of these regions
    - create joint umap
    - link peaks to genes
    - plot expression of peak-linked genes


# Resources
Methods and applications for single-cell and spatial multi-omics: https://www.nature.com/articles/s41576-023-00580-2

Into the multiverse: advances in single-cell multiomic profiling: https://www.cell.com/trends/genetics/fulltext/S0168-9525(22)00077-4

https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression

https://stuartlab.org/signac/

Merging ATAC datasets: https://stuartlab.org/signac/articles/merging.html